contrib/gitlab/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
+	sourcesv1alpha1 "github.com/knative/eventing-sources/contrib/gitlab/pkg/apis/sources/v1alpha1"
+	"github.com/knative/eventing-sources/pkg/kncloudevents"
+	gl "gopkg.in/go-playground/webhooks.v5/gitlab"
+)
+
+const (
+	GLHeaderEvent    = "GitLab-Event"
+	GLHeaderDelivery = "GitLab-Delivery"
+)
+
+// Adapter converts incoming GitLab webhook events to CloudEvents
+type Adapter struct {
+	client client.Client
+}
+
+// New creates an adapter to convert incoming GitLab webhook events to CloudEvents and
+// then sends them to the specified Sink
+func New(sinkURI string) (*Adapter, error) {
+	a := new(Adapter)
+	var err error
+	a.client, err = kncloudevents.NewDefaultClient(sinkURI)
+	if err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+// HandleEvent is invoked whenever an event comes in from GitLab
+func (a *Adapter) HandleEvent(payload interface{}, header http.Header) {
+	hdr := http.Header(header)
+	err := a.handleEvent(payload, hdr)
+	if err != nil {
+		log.Printf("unexpected error handling GitLab event: %s", err)
+	}
+}
+
+func (a *Adapter) handleEvent(payload interface{}, hdr http.Header) error {
+	gitLabEventType := hdr.Get("X-" + GLHeaderEvent)
+	eventID := hdr.Get("X-" + GLHeaderDelivery)
+	extensions := map[string]interface{}{
+		GLHeaderEvent:    gitLabEventType,
+		GLHeaderDelivery: eventID,
+	}
+
+	log.Printf("Handling %s", gitLabEventType)
+
+	cloudEventType := fmt.Sprintf("%s.%s", sourcesv1alpha1.GitLabSourceEventPrefix, gitLabEventType)
+	source, err := sourceFromGitLabEvent(gl.Event(gitLabEventType), payload)
+	if err != nil {
+		return err
+	}
+
+	event := cloudevents.Event{
+		Context: cloudevents.EventContextV02{
+			ID:         eventID,
+			Type:       cloudEventType,
+			Source:     *source,
+			Extensions: extensions,
+		}.AsV02(),
+		Data: payload,
+	}
+	_, err = a.client.Send(context.TODO(), event)
+	return err
+}
+
+func sourceFromGitLabEvent(gitLabEvent gl.Event, payload interface{}) (*types.URLRef, error) {
+	var url string
+	switch gitLabEvent {
+	case gl.PushEvents:
+		pe := payload.(gl.PushEventPayload)
+		url = pe.Repository.URL
+	case gl.TagEvents:
+		te := payload.(gl.TagEventPayload)
+		url = te.Repository.URL
+	case gl.IssuesEvents:
+		ie := payload.(gl.IssueEventPayload)
+		url = ie.ObjectAttributes.URL
+	case gl.ConfidentialIssuesEvents:
+		cie := payload.(gl.ConfidentialIssueEventPayload)
+		url = cie.ObjectAttributes.URL
+	case gl.CommentEvents:
+		ce := payload.(gl.CommentEventPayload)
+		url = ce.ObjectAttributes.URL
+	case gl.MergeRequestEvents:
+		mre := payload.(gl.MergeRequestEventPayload)
+		url = mre.ObjectAttributes.URL
+	case gl.WikiPageEvents:
+		wpe := payload.(gl.WikiPageEventPayload)
+		url = wpe.ObjectAttributes.URL
+	case gl.PipelineEvents:
+		pe := payload.(gl.PipelineEventPayload)
+		url = pe.ObjectAttributes.URL
+	case gl.BuildEvents:
+		be := payload.(gl.BuildEventPayload)
+		url = be.Repository.URL
```

Comment:
That is really confusing that build events do not have `object_attributes` in the payload, but pipelines events do. This is not consistent across events, perhaps we should open an issue in GitLab about that.

Reply:
Agree with you